haigha/connection.py

```diff
@@ -404,7 +404,10 @@ def read_frames(self):
         # Send a heartbeat (if needed)
         self._channels[0].send_heartbeat()
 
-        data = self._transport.read(self._heartbeat)
+        # Wait for 2 heartbeat intervals before giving up. See AMQP 4.2.7:
+        # "If a peer detects no incoming traffic (i.e. received octets) for two heartbeat intervals or longer,
+        # it should close the connection"
+        data = self._transport.read(self._heartbeat*2)
```

Comment:
It has the logic, but it never does so voluntarily - only when we go into `read_frames`. It read frames waits for too long, a heartbeat will arrive from RabbitMQ which will be replied to - which is OK.
A side-effect to my change is doubling the period between client-sent heartbeats, but that's fine as RabbitMQ will send its own heartbeats which will be replied to.